File: Vrml2Sourcebook/Chapter18TextureMapping/Figure18_21TextureCoordinatePizzaSlice.py
####################################################################################################
#
# Invoking X3D model self-test:
#
#   $ python Figure18_21TextureCoordinatePizzaSlice.py
#
# Python package x3d.py package is available on PyPI for import.
#   This approach simplifies Python X3D deployment and use.
#   https://pypi.org/project/x3d
#
# Installation:
#       pip install x3d
# or
#       python -m pip install x3d
#
# Developer options for loading x3d package in other Python programs:
#
#    from x3d import *  # preferred approach, terser source that avoids x3d.* class prefixes
#
# or
#    import x3d         # traditional way to subclass x3d package, all classes require x3d.* prefix,
#                       # but python source is very verbose, for example x3d.Material x3d.Shape etc.
#                       # X3dToPython.xslt stylesheet insertPackagePrefix=true supports this option.
#
####################################################################################################

from x3d import *

newModel=X3D(profile='Interchange',version='3.0',
  head=head(
    children=[
    meta(content='Figure18_21TextureCoordinatePizzaSlice.x3d',name='title'),
    meta(content='Figure 18.21, The VRML 2.0 Sourcebook, Copyright [1997] By Andrea L. Ames, David R. Nadeau, and John L. Moreland',name='creator'),
    meta(content='http://www.wiley.com/legacy/compbooks/vrml2sbk/ch18/18fig21.htm',name='reference'),
    meta(content='Don Brutzman',name='translator'),
    meta(content='4 September 2000',name='created'),
    meta(content='20 October 2019',name='modified'),
    meta(content='A pizza and slice. Also see Figure 18.20b and Figure 18.21 for comparison.',name='description'),
    meta(content='Figure18_20bTextureCoordinatePizzaSlice.x3d',name='reference'),
    meta(content='https://www.web3d.org/x3d/content/examples/Vrml2Sourcebook/Chapter18TextureMapping/Figure18_21TextureCoordinatePizzaSlice.x3d',name='identifier'),
    meta(content='X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit',name='generator'),
    meta(content='../../license.html',name='license')]),
  Scene=Scene(
    children=[
    WorldInfo(title='Figure18_21TextureCoordinatePizzaSlice.x3d'),
    Viewpoint(description='Hey pepperoni',orientation=(1,0,0,-0.758),position=(0,2,2)),
    Shape(
      appearance=Appearance(
        texture=ImageTexture(url=["pizza.jpg","https://www.web3d.org/x3d/content/examples/Vrml2Sourcebook/Chapter18TextureMapping/pizza.jpg"])),
      geometry=IndexedFaceSet(coordIndex=[0,1,2,3,4,5,-1,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35],solid=False,texCoordIndex=[0,28,29,30,31,32,-1,0,32,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28],
        coord=Coordinate(point=[(0.50,0,0.50),(0.88,0,1.42),(1.06,0,1.33),(1.21,0,1.21),(1.33,0,1.06),(1.42,0,0.88),(0.00,0,0.00),(0.92,0,0.38),(0.98,0,0.20),(1.00,0,0.00),(0.98,0,-0.20),(0.92,0,-0.38),(0.83,0,-0.56),(0.71,0,-0.71),(0.56,0,-0.83),(0.38,0,-0.92),(0.20,0,-0.98),(0.00,0,-1.00),(-0.20,0,-0.98),(-0.38,0,-0.92),(-0.56,0,-0.83),(-0.71,0,-0.71),(-0.83,0,-0.56),(-0.92,0,-0.38),(-0.98,0,-0.20),(-1.00,0,0.00),(-0.98,0,0.20),(-0.92,0,0.38),(-0.83,0,0.56),(-0.71,0,0.71),(-0.56,0,0.83),(-0.38,0,0.92),(-0.20,0,0.98),(0.00,0,1.00),(0.20,0,0.98),(0.38,0,0.92)]),
        texCoord=TextureCoordinate(point=[(0.50,0.50),(0.96,0.41),(0.97,0.50),(0.96,0.59),(0.93,0.68),(0.89,0.76),(0.83,0.83),(0.76,0.89),(0.68,0.93),(0.59,0.96),(0.50,0.97),(0.41,0.96),(0.32,0.93),(0.24,0.89),(0.17,0.83),(0.11,0.76),(0.07,0.68),(0.04,0.59),(0.03,0.50),(0.04,0.41),(0.07,0.32),(0.11,0.24),(0.17,0.17),(0.24,0.11),(0.32,0.07),(0.41,0.04),(0.50,0.03),(0.59,0.04),(0.68,0.07),(0.76,0.11),(0.83,0.17),(0.89,0.24),(0.93,0.32)])))])
) # X3D model complete

####################################################################################################
# Self-test diagnostics
####################################################################################################

print('Self-test diagnostics for Figure18_21TextureCoordinatePizzaSlice.py:')
if        metaDiagnostics(newModel): # built-in utility method in X3D class
    print(metaDiagnostics(newModel)) # display meta info, hint, warning, error, TODO values in this model
# print('check newModel.XML() serialization...')
newModelXML= newModel.XML() # test export method XML() for exceptions during export
newModel.XMLvalidate()
# print(newModelXML) # diagnostic

try:
#   print('check newModel.VRML() serialization...')
    newModelVRML=newModel.VRML() # test export method VRML() for exceptions during export
    # print(prependLineNumbers(newModelVRML)) # debug
    print("Python-to-VRML export of VRML output successful", flush=True)
except Exception as err: # usually BaseException
    # https://stackoverflow.com/questions/18176602/how-to-get-the-name-of-an-exception-that-was-caught-in-python
    print("*** Python-to-VRML export of VRML output failed:", type(err).__name__, err)
    if newModelVRML: # may have failed to generate
        print(prependLineNumbers(newModelVRML, err.lineno))

try:
#   print('check newModel.JSON() serialization...')
    newModelJSON=newModel.JSON() # test export method JSON() for exceptions during export
#   print(prependLineNumbers(newModelJSON)) # debug
    print("Python-to-JSON export of JSON output successful (under development)")
except Exception as err: # usually SyntaxError
    print("*** Python-to-JSON export of JSON output failed:", type(err).__name__, err)
    if newModelJSON: # may have failed to generate
        print(prependLineNumbers(newModelJSON,err.lineno))

print("python Figure18_21TextureCoordinatePizzaSlice.py load and self-test diagnostics complete.")

File: Vrml2Sourcebook/Chapter18TextureMapping/Figure18_21TextureCoordinatePizzaSlice.java
package Vrml2Sourcebook.Chapter18TextureMapping;

/*
Copyright (c) 1995-2023 held by the author(s).  All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer
      in the documentation and/or other materials provided with the
      distribution.
    * Neither the name of the Web3D Consortium (https://www.web3d.org)
      nor the names of its contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
*/

import org.web3d.x3d.jsail.Core.*;
import org.web3d.x3d.jsail.fields.*;
import org.web3d.x3d.jsail.Geometry3D.*;
import org.web3d.x3d.jsail.Navigation.*;
import org.web3d.x3d.jsail.Rendering.*;
import org.web3d.x3d.jsail.Shape.*;
import org.web3d.x3d.jsail.Texturing.*;

// Javadoc metadata annotations follow, see below for X3DJSAIL Java source code.
/**
 * <p> A pizza and slice. Also see Figure 18.20b and Figure 18.21 for comparison. </p>
 <p> Related links: <a href="../../../Chapter18TextureMapping/Figure18_21TextureCoordinatePizzaSlice.java">Figure18_21TextureCoordinatePizzaSlice.java</a> source, <a href="../../../Chapter18TextureMapping/Figure18_21TextureCoordinatePizzaSliceIndex.html" target="_top">Figure18_21TextureCoordinatePizzaSlice catalog page</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">X3D Resources</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html" target="_blank">X3D Scene Authoring Hints</a>, and <a href="https://www.web3d.org/x3d/content/X3dTooltips.html" target="_blank">X3D Tooltips</a>. </p>
	<table style="color:black; border:0px solid; border-spacing:10px 0px;">
        <caption>Scene Meta Information</caption>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center; padding:10px 0px;"><i>meta tags</i></td>
			<td style="text-align:left;   padding:10px 0px;">&nbsp; Document Metadata </td>
		</tr>

		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> title </i> </td>
			<td> <a href="../../../Chapter18TextureMapping/Figure18_21TextureCoordinatePizzaSlice.x3d">Figure18_21TextureCoordinatePizzaSlice.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> creator </i> </td>
			<td> Figure 18.21, The VRML 2.0 Sourcebook, Copyright [1997] By Andrea L. Ames, David R. Nadeau, and John L. Moreland </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> reference </i> </td>
			<td> <a href="http://www.wiley.com/legacy/compbooks/vrml2sbk/ch18/18fig21.htm" target="_blank">http://www.wiley.com/legacy/compbooks/vrml2sbk/ch18/18fig21.htm</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> translator </i> </td>
			<td> Don Brutzman </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> created </i> </td>
			<td> 4 September 2000 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> modified </i> </td>
			<td> 20 October 2019 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> description </i> </td>
			<td> A pizza and slice. Also see Figure 18.20b and Figure 18.21 for comparison. </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> reference </i> </td>
			<td> <a href="../../../Chapter18TextureMapping/Figure18_20bTextureCoordinatePizzaSlice.x3d">Figure18_20bTextureCoordinatePizzaSlice.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> identifier </i> </td>
			<td> <a href="https://www.web3d.org/x3d/content/examples/Vrml2Sourcebook/Chapter18TextureMapping/Figure18_21TextureCoordinatePizzaSlice.x3d" target="_blank">https://www.web3d.org/x3d/content/examples/Vrml2Sourcebook/Chapter18TextureMapping/Figure18_21TextureCoordinatePizzaSlice.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> generator </i> </td>
			<td> X3D-Edit 3.3, <a href="https://savage.nps.edu/X3D-Edit" target="_blank">https://savage.nps.edu/X3D-Edit</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> license </i> </td>
			<td> <a href="../../../Chapter18TextureMapping/../../license.html">../../license.html</a> </td>
		</tr>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center;" colspan="2">  &nbsp; </td>
		</tr>
	</table>

	<p>
		This program uses the
		<a href="https://www.web3d.org/specifications/java/X3DJSAIL.html" target="_blank">X3D Java Scene Access Interface Library (X3DJSAIL)</a>.
		It has been produced using the 
		<a href="https://www.web3d.org/x3d/stylesheets/X3dToJava.xslt" target="_blank">X3dToJava.xslt</a>
		stylesheet
	       (<a href="https://sourceforge.net/p/x3d/code/HEAD/tree/www.web3d.org/x3d/stylesheets/X3dToJava.xslt" target="_blank">version&amp;nbsp;control</a>)
                is used to create Java source code from an original <code>.x3d</code> model.
	</p>

	* @author Figure 18.21, The VRML 2.0 Sourcebook, Copyright [1997] By Andrea L. Ames, David R. Nadeau, and John L. Moreland
 */

public class Figure18_21TextureCoordinatePizzaSlice
{
	/** Default constructor to create this object. */
	public Figure18_21TextureCoordinatePizzaSlice ()
	{
	  initialize();
	}

	/** Create and initialize the X3D model for this object. */
	public final void initialize()
	{
            try { // catch-all
  x3dModel = new X3D().setProfile(X3D.PROFILE_INTERCHANGE).setVersion(X3D.VERSION_3_0)
  .setHead(new head()
    .addMeta(new meta().setName(meta.NAME_TITLE      ).setContent("Figure18_21TextureCoordinatePizzaSlice.x3d"))
    .addMeta(new meta().setName(meta.NAME_CREATOR    ).setContent("Figure 18.21, The VRML 2.0 Sourcebook, Copyright [1997] By Andrea L. Ames, David R. Nadeau, and John L. Moreland"))
    .addMeta(new meta().setName(meta.NAME_REFERENCE  ).setContent("http://www.wiley.com/legacy/compbooks/vrml2sbk/ch18/18fig21.htm"))
    .addMeta(new meta().setName(meta.NAME_TRANSLATOR ).setContent("Don Brutzman"))
    .addMeta(new meta().setName(meta.NAME_CREATED    ).setContent("4 September 2000"))
    .addMeta(new meta().setName(meta.NAME_MODIFIED   ).setContent("20 October 2019"))
    .addMeta(new meta().setName(meta.NAME_DESCRIPTION).setContent("A pizza and slice. Also see Figure 18.20b and Figure 18.21 for comparison."))
    .addMeta(new meta().setName(meta.NAME_REFERENCE  ).setContent("Figure18_20bTextureCoordinatePizzaSlice.x3d"))
    .addMeta(new meta().setName(meta.NAME_IDENTIFIER ).setContent("https://www.web3d.org/x3d/content/examples/Vrml2Sourcebook/Chapter18TextureMapping/Figure18_21TextureCoordinatePizzaSlice.x3d"))
    .addMeta(new meta().setName(meta.NAME_GENERATOR  ).setContent("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"))
    .addMeta(new meta().setName(meta.NAME_LICENSE    ).setContent("../../license.html")))
  .setScene(new Scene()
    .addChild(new WorldInfo().setTitle("Figure18_21TextureCoordinatePizzaSlice.x3d"))
    .addChild(new Viewpoint().setDescription("Hey pepperoni").setOrientation(1.0,0.0,0.0,-0.758).setPosition(0.0,2.0,2.0))
    .addChild(new Shape()
      .setAppearance(new Appearance()
        .setTexture(new ImageTexture().setUrl(new String[] {"pizza.jpg","https://www.web3d.org/x3d/content/examples/Vrml2Sourcebook/Chapter18TextureMapping/pizza.jpg"})))
      .setGeometry(new IndexedFaceSet().setSolid(false).setCoordIndex(new int[] {0,1,2,3,4,5,-1,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35}).setTexCoordIndex(new int[] {0,28,29,30,31,32,-1,0,32,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28})
        .setCoord(new Coordinate().setPoint(new MFVec3f(new double[] {0.50,0.0,0.50,0.88,0.0,1.42,1.06,0.0,1.33,1.21,0.0,1.21,1.33,0.0,1.06,1.42,0.0,0.88,0.00,0.0,0.00,0.92,0.0,0.38,0.98,0.0,0.20,1.00,0.0,0.00,0.98,0.0,-0.20,0.92,0.0,-0.38,0.83,0.0,-0.56,0.71,0.0,-0.71,0.56,0.0,-0.83,0.38,0.0,-0.92,0.20,0.0,-0.98,0.00,0.0,-1.00,-0.20,0.0,-0.98,-0.38,0.0,-0.92,-0.56,0.0,-0.83,-0.71,0.0,-0.71,-0.83,0.0,-0.56,-0.92,0.0,-0.38,-0.98,0.0,-0.20,-1.00,0.0,0.00,-0.98,0.0,0.20,-0.92,0.0,0.38,-0.83,0.0,0.56,-0.71,0.0,0.71,-0.56,0.0,0.83,-0.38,0.0,0.92,-0.20,0.0,0.98,0.00,0.0,1.00,0.20,0.0,0.98,0.38,0.0,0.92})))
        .setTexCoord(new TextureCoordinate().setPoint(new MFVec2f(new double[] {0.50,0.50,0.96,0.41,0.97,0.50,0.96,0.59,0.93,0.68,0.89,0.76,0.83,0.83,0.76,0.89,0.68,0.93,0.59,0.96,0.50,0.97,0.41,0.96,0.32,0.93,0.24,0.89,0.17,0.83,0.11,0.76,0.07,0.68,0.04,0.59,0.03,0.50,0.04,0.41,0.07,0.32,0.11,0.24,0.17,0.17,0.24,0.11,0.32,0.07,0.41,0.04,0.50,0.03,0.59,0.04,0.68,0.07,0.76,0.11,0.83,0.17,0.89,0.24,0.93,0.32}))))));
            }
            catch (Exception ex)
            {       
                System.err.println ("*** Further hints on X3DJSAIL errors and exceptions at");
                System.err.println ("*** https://www.web3d.org/specifications/java/X3DJSAIL.html");
                throw (ex);
            }
	}
	// end of initialize() method

	/** The initialized model object, created within initialize() method. */
	private X3D x3dModel;

	/** 
	 * Provide a 
	 * <a href="https://dzone.com/articles/java-copy-shallow-vs-deep-in-which-you-will-swim" target="_blank">shallow copy</a>
	 * of the X3D model.
	 * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/Core/X3D.html">X3D</a>
	 * @return Figure18_21TextureCoordinatePizzaSlice model
	 */
	public X3D getX3dModel()
	{	  
		return x3dModel;
	}
	   
    /** 
     * Default main() method provided for test purposes, uses CommandLine to set global ConfigurationProperties for this object.
     * @param args array of input parameters, provided as arguments
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/Core/X3D.html#handleArguments-java.lang.String:A-">X3D.handleArguments(args)</a>
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/Core/X3D.html#validationReport--">X3D.validationReport()</a>
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/CommandLine.html">CommandLine</a>
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/CommandLine.html#USAGE">CommandLine.USAGE</a>
     * @see <a href="https://www.web3d.org/specifications/java/javadoc/org/web3d/x3d/jsail/ConfigurationProperties.html">ConfigurationProperties</a>
     */
    public static void main(String args[])
    {
        System.out.println("Build this X3D model, showing validation diagnostics...");
        X3D thisExampleX3dModel = new Figure18_21TextureCoordinatePizzaSlice().getX3dModel();
//      System.out.println("X3D model construction complete.");
	
        // next handle command line arguments
        boolean hasArguments = (args != null) && (args.length > 0);
        boolean validate = true; // default
        boolean argumentsLoadNewModel = false;
        String  fileName = new String();

        if (args != null)
        {
                for (String arg : args)
                {
                        if (arg.toLowerCase().startsWith("-v") || arg.toLowerCase().contains("validate"))
                        {
                                validate = true; // making sure
                        }
                        if (arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_X3D) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_CLASSICVRML) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_X3DB) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_VRML97) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_EXI) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_GZIP) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_ZIP) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_HTML) ||
                                arg.toLowerCase().endsWith(X3D.FILE_EXTENSION_XHTML))
                        {
                                argumentsLoadNewModel = true;
                                fileName = arg;
                        }
                }
        }
        if      (argumentsLoadNewModel)
                System.out.println("WARNING: \"Vrml2Sourcebook.Chapter18TextureMapping.Figure18_21TextureCoordinatePizzaSlice\" model invocation is attempting to load file \"" + fileName + "\" instead of simply validating itself... file loading ignored.");
        else if (hasArguments) // if no arguments provided, this method produces usage warning
                thisExampleX3dModel.handleArguments(args);
	
        if (validate)
        {
            //  System.out.println("--- TODO fix duplicated outputs ---"); // omit when duplicated outputs problem is solved/refactored
		String validationResults = thisExampleX3dModel.validationReport();
            //  System.out.println("-----------------------------------"); // omit when duplicated outputs problem is solved/refactored
                System.out.print("Vrml2Sourcebook.Chapter18TextureMapping.Figure18_21TextureCoordinatePizzaSlice self-validation test results: ");
                if (!validationResults.equals("success"))
                    System.out.println();
                System.out.println(validationResults.trim());
        }
    }
}
